src/dhtproto/client/legacy/internal/helper/Handshake.d

```diff
@@ -0,0 +1,241 @@
+            been reached
+
+        ***********************************************************************/
+
+        private void allNodesConnected ()
+        {
+            this.outer.log.info("All DHT nodes connected!");
+            this.all_connected_event.trigger();
+        }
+
+        /***********************************************************************
+
+            Runs the task
+
+        ***********************************************************************/
+
+        protected override void run ()
+        {
+            this.all_connected_event.wait();
+        }
+    }
+
+
+    /***************************************************************************
+
+        Task that suspends until at least one node is connected
+
+    ***************************************************************************/
+
+    private class OneDhtNodeConnected : Task
+    {
+        import swarm.Const : NodeItem;
+
+        /// Used to manage suspend/resume of this task
+        private TaskEvent node_connected_event;
+
+        /***********************************************************************
+
+            Delegate to be called by the handshake each time an individual
+            node connects
+
+            Params:
+                item = data on the address and port of the connected node
+
+        ***********************************************************************/
+
+        private void oneNodeConnected (NodeItem item)
+        {
+            this.outer.log.trace("Connected to DHT node {}:{}",
+                                 item.Address, item.Port);
+            this.node_connected_event.trigger();
+        }
+
+        /***********************************************************************
+
+            Runs the task
+
+        ***********************************************************************/
+
+        protected override void run ()
+        {
+            this.node_connected_event.wait();
+        }
+    }
+
+
+    /// Logger for the handshake
+    private Logger log;
+
+
+    /// Instance of task that suspends until all nodes are connected
+    private AllDhtNodesConnected all_nodes_connected;
+
+
+    /// Instance of task that suspends until at least one node is connected
+    private OneDhtNodeConnected one_node_connected;
+
+
+    /// Instance of class that launches the actual handshake, retrying when
+    /// it fails until all nodes are connected
+    private RetryHandshake retry_handshake;
+
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            dht = DHT client instance via which to perform the handshake
+            retry_delay_seconds = time to wait (in seconds) before retrying
+                failed handshake attempts
+
+    ***************************************************************************/
+
+    public this (DhtClient dht, ulong retry_delay_seconds)
```

Comment:
Minor note: `RetryHandshake`'s `wait_time` parameter is `size_t`.  Perhaps we should use `uint` here to guarantee it is compatible?  (Yes, we only support 64-bit, but this is low-hanging fruit to avoid incompatibility with 32-bit.)

Reply:
We should also use `size_t`.

Reply:
Why?  It's a weird type to use for a value in seconds.

Reply:
To match upstream definition and avoid 32-bit incompatibility.

Reply:
`uint` will be compatible with the `size_t` parameter of `RetryHandshake`, as will `int` with a `verify(retry_delay_seconds > 0)` under the hood (@nemanja-boric-sociomantic expressed a preference for a signed value).